dev-tools/gen-target-info/src/main.rs

```diff
@@ -1,52 +1,73 @@
-use gen_target_info::{get_target_specs_from_json, write_target_tuple_mapping, RustcTargetSpecs};
-use std::{collections::BTreeMap, fs::File, io::Write as _};
+use std::io::Write as _;
+use std::{fs::File, io::BufRead};
+
+use gen_target_info::{
+    get_target_spec_from_msrv, get_target_specs_from_json, get_targets_msrv, RustcTargetSpecs,
+};
 
 const PRELUDE: &str = r#"//! This file is generated code. Please edit the generator
 //! in dev-tools/gen-target-info if you need to make changes.
 
 "#;
 
-fn generate_riscv_arch_mapping(f: &mut File, target_specs: &RustcTargetSpecs) {
-    let riscv_target_mapping = target_specs
-        .0
-        .iter()
-        .filter_map(|(target, target_spec)| {
-            let arch = target.split_once('-').unwrap().0;
-            (arch.contains("riscv") && arch != target_spec.arch)
-                .then_some((arch, &*target_spec.arch))
-        })
-        .collect::<BTreeMap<_, _>>();
-    write_target_tuple_mapping(f, "RISCV_ARCH_MAPPING", &riscv_target_mapping);
-}
+fn generate_target_mapping(f: &mut File, target_specs: &RustcTargetSpecs) -> std::io::Result<()> {
+    writeln!(f, "use super::Target;")?;
+    writeln!(f)?;
+    writeln!(
+        f,
+        "pub(crate) fn get(target_triple: &str) -> Option<Target> {{"
+    )?;
+    writeln!(f, "    Some(match target_triple {{")?;
+
+    for (triple, spec) in &target_specs.0 {
+        let full_arch = triple.split_once('-').unwrap().0;
+        let arch = &spec.arch;
+        let vendor = spec.vendor.as_deref().unwrap_or("unknown");
+        let os = spec.os.as_deref().unwrap_or("none");
+        let env = spec.env.as_deref().unwrap_or("");
+        let abi = spec.abi.as_deref().unwrap_or("");
+
+        writeln!(f, "        {triple:?} => Target {{")?;
+        writeln!(f, "            full_arch: {full_arch:?}.into(),")?;
+        writeln!(f, "            arch: {arch:?}.into(),")?;
+        writeln!(f, "            vendor: {vendor:?}.into(),")?;
+        writeln!(f, "            os: {os:?}.into(),")?;
+        writeln!(f, "            env: {env:?}.into(),")?;
+        writeln!(f, "            abi: {abi:?}.into(),")?;
+        writeln!(f, "        }},")?;
```

Comment:
Would it possible to have an array of `[(&'static str, Target); N]`?

I think writing a large match here is less readable than having a sorted array.

Reply:
I also suspect that a match would generate more code than a linear/binary search in sorted array, thus taking longer to compile.

Matching would basically be lots of `if`s, the compiler might also optimize the `if` into a jump table, but that would still mean a lot of `Target` creation at runtime with OOM panic handling blocks.

Reply:
Huh, I seemed to remember `Cow` not really being usable in `const`s because they have drop glue, but apparently I remembered wrong, and I think it's doable.